server/routes/auth.ts

```diff
@@ -76,470 +48,108 @@ passport.deserializeUser<IUser, string>((id, done) => {
-	profileFields: ["id", "displayName", "email"],
-	passReqToCallback: true
-});
+		user.accountConfirmed = true;
 
-const PBKDF2_ROUNDS: number = 300000;
-passport.use(new LocalStrategy({
-	usernameField: "email",
-	passwordField: "password",
-	passReqToCallback: true
-}, async (request, email, password, done) => {
-	let user = await User.findOne({ "email": email });
-	if (user && request.path.match(/\/signup$/i)) {
-		done(null, false, { "message": "That email address is already in use" });
-	}
-	else if (user && !user.localData!.hash && (user.githubData!.id || user.googleData!.id || user.facebookData!.id)) {
-		done(null, false, { "message": "Please log back in with GitHub, Google, or Facebook" });
-	}
-	else if (!user || !user.localData) {
-		// User hasn't signed up yet
-		if (!request.path.match(/\/signup$/i)) {
-			// Only create the user when targeting /signup
-			done(null, false, { "message": "Incorrect email or password" });
-			return;
-		}
-		let name: string = request.body.name || "";
-		name = name.trim();
-		if (!name || !email || !password) {
-			done(null, false, { "message": "Missing email, name, or password" });
-			return;
-		}
-		let isAdmin = false; // Only set this after they have verified their email
-		let salt = crypto.randomBytes(32);
-		let hash = await pbkdf2Async(password, salt, PBKDF2_ROUNDS);
-		user = new User({
-			"uuid": uuid(),
-			"email": email,
-			"name": request.body.name,
-			"verifiedEmail": false,
-
-			"localData": {
-				"hash": hash.toString("hex"),
-				"salt": salt.toString("hex"),
-				"verificationCode": crypto.randomBytes(32).toString("hex")
-			},
-			"githubData": {},
-			"googleData": {},
-			"facebookData": {},
-
-			"applied": false,
-			"accepted": false,
-			"preConfirmEmailSent": false,
-			"confirmed": false,
-			"applicationData": [],
-			"applicationStartTime": undefined,
-			"applicationSubmitTime": undefined,
-
-			"admin": isAdmin
-		});
 		try {
 			await user.save();
-			trackEvent("created account", request, email);
-		}
-		catch (err) {
-			done(err);
-			return;
-		}
-		// Send verification email (hostname validated by previous middleware)
-		let link = createLink(request, `/auth/verify/${user.localData!.verificationCode}`);
-		let markdown =
-`Hi {{name}},
-
-Thanks for signing up for ${config.eventName}! To verify your email, please [click here](${link}).
-
-Sincerely,
-
-The ${config.eventName} Team.`;
-		try {
-			await sendMailAsync({
-				from: config.email.from,
-				to: email,
-				subject: `[${config.eventName}] - Verify your email`,
-				html: await renderEmailHTML(markdown, user),
-				text: await renderEmailText(markdown, user)
-			});
-		}
-		catch (err) {
-			done(err);
-			return;
-		}
-		done(null, user);
-	}
-	else {
-		// Log the user in
-		let hash = await pbkdf2Async(password, Buffer.from(user.localData.salt, "hex"), PBKDF2_ROUNDS);
-		if (hash.toString("hex") === user.localData.hash) {
-			if (user.verifiedEmail) {
-				done(null, user);
+			if (email) {
```

Comment:
Yes because `email != user.email`. The email in this case is passed along by the account confirmation form and doesn't exist if the email can't be changed (like when from OAuth providers). Still, I agree it's a bit confusing so I changed to send an email if they don't have a verified email and no verification email has been sent yet.